### PR TITLE
Create test-plan/overview for constructor tear-off

### PR DESCRIPTION
**Pull request by lrhn**
Not sure exactly what I'm aiming for here.
Tried to make an overview of the new features, and things that might be worth testing.
That's a potentially big matrix of combinations.

### CODE REVIEW

File: accepted/future-releases/constructor-tearoffs/test-plan.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+The Dart constructor tear-off feature contains multiple independent language changes. Each should be tested, as should their combination with each other and with other relevant language features.
+
+The features can be summarized as:
+
+* Named constructor tear-offs
+  * Expressions of the for <code>*C*.*name*</code>. or <code>*C*\<*typeArgs*>.*name*</code>, with type arguments potentially being inferred, evaluate to functions.
+  * Preserves identity as specified.
+* Unnamed constructor syntax alternative
+  * Everywhere you can currently reference an unnamed constructor, you can also refer to it as <code>*C*.new</code>.
```
Comment (leafpetersen) [inline accepted/future-releases/constructor-tearoffs/test-plan.md:11]: And declare?

File: accepted/future-releases/constructor-tearoffs/test-plan.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+* Factory or not
+* Redirecting or not
+
+The possible sources of tear-offs:
+
+* From a class declaring the constructor.
+* From a mixin-application which introduces a forwarding constructor.
+* Any of those through an alias which is:
+  * Not generic.
+  * A proper rename (static type may not match runtime type, static type is based on alias type argument bounds, runtime type is based on class type argument bounds).
```
Comment (leafpetersen) [inline accepted/future-releases/constructor-tearoffs/test-plan.md:46]: Specifically need to check that tearoffs from proper renames of generics aren't treated as instantiate to bounds, but instead are subject to normal instance creation inference at a call site (e.g. `C<int> x = TypeDef.name(3)` should use the context type to fill in the missing type argument to `TypeDef`).

Comment (leafpetersen) [reply accepted/future-releases/constructor-tearoffs/test-plan.md:46]: Also should ensure that bounds are checked.

File: accepted/future-releases/constructor-tearoffs/test-plan.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+### Incorrect usage
+
+* Does not work for function values.
+  * Including torn off methods and constructors.
+* Includes function-typed getters.
+* Does not work for `call` methods of function values.
+* Does not work for *constructors* (`class C<T> { C.name();}` does not allow `C.name<int>`).
+* Non-constant instantiation in constant context.
+
```
Comment (leafpetersen) [inline accepted/future-releases/constructor-tearoffs/test-plan.md:125]: Check that type bounds are checked.

File: accepted/future-releases/constructor-tearoffs/test-plan.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+### Incorrect usage
+
+* Non-constant instantiation in constant context.
+  * Even if type parameter of type alias is not used in result.
+
+* Cannot call static members on instantiated type literals. (That's a constructor reference, if anything).
+  * <code>*C*\<*typeArgs*>.anything</code> does not create a type literal.
+  * <code>*C*\<*typeArgs*>..anything</code> *probably does* (<code>*C*..toString()</code> works today, invokes `toString` of `Type` object).
+  * <code>*C*\<*typeArgs*>?.anything</code> also works (possibly with a warning about the `?` being unnecessary).
+
```
Comment (leafpetersen) [inline accepted/future-releases/constructor-tearoffs/test-plan.md:154]: Check that type bounds are checked.

**Review by leafpetersen — APPROVED** [review]
Nice, thanks!  Do you plan to update this with the "covered by test" entries, or do you think it would be better to move those to tracking issues?

**Comment by mit-mit** [comment]
@lrhn shall we merge this?

**Pull request merged**